### PR TITLE
sentence piece tokenizer support for TokenizerInfo

### PR DESCRIPTION
**Pull request by zanderjiang**

### CODE REVIEW

File: python/xgrammar/tokenizer_info.py [hunk 1]
```diff
@@ -39,6 +41,35 @@ class VocabType(Enum):
     BYTE_FALLBACK = "BYTE_FALLBACK"
     BYTE_LEVEL = "BYTE_LEVEL"
 
+def is_tiktoken_tokenizer(tokenizer: PreTrainedTokenizerBase) -> bool:
```
Comment (Ubospica) [inline python/xgrammar/tokenizer_info.py]: Use underscore to hide internal functions
```suggestion
def _is_tiktoken_tokenizer(tokenizer: PreTrainedTokenizerBase) -> bool:
```

File: python/xgrammar/tokenizer_info.py [hunk 2]
```diff
@@ -39,6 +41,35 @@ class VocabType(Enum):
...
+    )
+    
+    filename_pattern = (
+        "vocab_file" in tokenizer.vocab_files_names and 
+        "tiktoken" in tokenizer.vocab_files_names["vocab_file"]
+    )
+    
+    return has_tiktoken_encoding or filename_pattern
+
+def is_sentencepiece_tokenizer(tokenizer: PreTrainedTokenizerBase) -> bool:
```
Comment (Ubospica) [inline python/xgrammar/tokenizer_info.py]: Ditto

File: python/xgrammar/tokenizer_info.py [hunk 3]
```diff
@@ -139,17 +170,28 @@ def from_huggingface(
...
+            vocab_dict = tokenizer.get_vocab()
         except AttributeError as e:
             msg = (
                 f"Cannot get the vocabulary of the tokenizer {type(tokenizer)}. The tokenizer "
                 "should have a get_vocab method."
             )
             raise ValueError(msg) from e
 
+        max_id = max(vocab_dict.values()) if vocab_dict else -1
+        min_vocab_size = max(len(vocab_dict), max_id + 1)
```
Comment (Ubospica) [inline python/xgrammar/tokenizer_info.py]: Rename to detected_vocab_size

**Review by Ubospica — COMMENTED** [review]
Looks great!

Could you leave a benchmark result in comments of this pr:
- The cpu you are using
- The model used (three kinds: hf, tiktoken, sp)
- TokenizerInfo Build time

We can merge it after addressing these issues.

**Comment by zanderjiang** [comment]
CPU: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz

Build-time tests:
"meta-llama/Meta-Llama-3-8B-Instruct" (hf)
build_time: 0.5978555679321289

"microsoft/Phi-3.5-mini-instruct" (hf)
build_time: 0.10898065567016602

"Qwen/Qwen-7B-Chat" (tiktoken)
build_time: 0.08476638793945312

"THUDM/glm-4-9b-chat" (tiktoken)
build_time: 0.11713862419128418

"THUDM/chatglm3-6b" (sp)
build_time: 0.1258544921875

**Review by Ubospica — APPROVED** [review]
Looks great. Thanks @zanderjiang!

**Pull request merged**